flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/AsyncCalcSplitRule.java

```diff
@@ -0,0 +1,194 @@
+    public static final RelOptRule REWRITE_PROJECT =
+            new RemoteCalcRewriteProjectionRule(ASYNC_CALL_FINDER);
+    public static final RelOptRule NESTED_SPLIT = new AsyncCalcSplitNestedRule(ASYNC_CALL_FINDER);
+    public static final RelOptRule ONE_PER_CALC_SPLIT =
+            new AsyncCalcSplitOnePerCalcRule(ASYNC_CALL_FINDER);
+
+    /**
+     * An Async implementation of {@link RemoteCalcCallFinder} which finds uses of
+     * AsyncScalarFunction.
+     */
+    public static class AsyncRemoteCalcCallFinder implements RemoteCalcCallFinder {
+        @Override
+        public boolean containsRemoteCall(RexNode node) {
+            return AsyncUtil.containsAsyncCall(node);
+        }
+
+        @Override
+        public boolean containsNonRemoteCall(RexNode node) {
+            return AsyncUtil.containsNonAsyncCall(node);
+        }
+
+        @Override
+        public boolean isRemoteCall(RexNode node) {
+            return AsyncUtil.isAsyncCall(node);
+        }
+
+        @Override
+        public boolean isNonRemoteCall(RexNode node) {
+            return AsyncUtil.isNonAsyncCall(node);
+        }
+    }
+
+    private static boolean hasNestedCalls(List<RexNode> projects) {
+        return projects.stream()
+                .filter(AsyncUtil::containsAsyncCall)
+                .filter(expr -> expr instanceof RexCall)
+                .map(expr -> (RexCall) expr)
+                .anyMatch(
+                        rexCall ->
+                                rexCall.getOperands().stream()
+                                        .anyMatch(AsyncUtil::containsAsyncCall));
+    }
+
+    /**
+     * Splits nested call <- asyncCall chains so that nothing is immediately waiting on an async
+     * call in a single calc.
+     */
+    public static class AsyncCalcSplitNestedRule extends RemoteCalcSplitRuleBase<Void> {
+
+        public AsyncCalcSplitNestedRule(RemoteCalcCallFinder callFinder) {
+            super("AsyncCalcSplitNestedRule", callFinder);
+        }
+
+        @Override
+        public boolean matches(RelOptRuleCall call) {
+            FlinkLogicalCalc calc = call.rel(0);
+
+            // Matches if we have nested remote calls
+            List<RexNode> projects =
+                    calc.getProgram().getProjectList().stream()
+                            .map(calc.getProgram()::expandLocalRef)
+                            .collect(Collectors.toList());
+            return hasNestedCalls(projects);
+        }
+
+        // We convert not on the outermost call, but anything within it
+        @Override
+        public boolean needConvert(RexProgram program, RexNode node, Option<Void> matchState) {
+            return node instanceof RexCall
+                    && !((RexCall) node)
+                            .getOperands().stream().anyMatch(callFinder()::containsRemoteCall);
+        }
+
+        @Override
+        public Tuple3<Option<RexNode>, Option<RexNode>, Seq<RexNode>> split(
```

Comment:
Never use Tuple3 or Pair. Use a little helper POJO so that the reader knows what the components are. Nobody forces to use getters or constructors.
```
class Pojo {
 T1 field;
 T2 field;
 T3 field;
}
```